### PR TITLE
feat(standardized form data): keep all columns and metrics

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

This PR introduced a new mechanism base on the [standardize form data](https://github.com/apache/superset/pull/20010). the metrics and columns will always keep its values until user remove it.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/2016594/173591690-589e04cf-c5a9-4260-b835-2802c562f4d0.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by kgabryje** [comment]
Thanks for the improvement, love how it works with the new viz switcher!
1 problem that I found while testing - it crashes when I change viz type from big number to pie (only 1 control crashes) and bar chart (the whole page crashes)

https://user-images.githubusercontent.com/15073128/173596160-f837bb59-b6dc-45ff-9d45-c9e6f1aa29fc.mov

**Comment by zhaoyongjie** [comment]
@kgabryje Thanks review! I have rebased the latest master and fix the bug. please review again.

**Comment by kgabryje** [comment]
@zhaoyongjie Tested again. It looks like Dimensions are not kept when I do `Open line chart -> switch to big number -> switch to bar chart`. However, it works correctly if I do `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to bar (dimensions are still kept)`.

Another example of flow working as expected: `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to bar (dimensions are still kept) -> switch to area chart (dimensions are still kept)`.
Another example of flow NOT working as expected: `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to area chart (dimensions are NOT kept)`

So it looks like dimensions are kept in target chart only if previous chart also had dimensions control.

**Comment by zhaoyongjie** [comment]
> @zhaoyongjie Tested again. It looks like Dimensions are not kept when I do `Open line chart -> switch to big number -> switch to bar chart`. However, it works correctly if I do `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to bar (dimensions are still kept)`.
> 
> Another example of flow working as expected: `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to bar (dimensions are still kept) -> switch to area chart (dimensions are still kept)`. Another example of flow NOT working as expected: `Open line chart -> switch to bar (dimensions are kept) -> switch to big number -> switch to area chart (dimensions are NOT kept)`
> 
> So it looks like dimensions are kept in target chart only if previous chart also had dimensions control.
> Screen.Recording.2022-06-15.at.13.43.40.mov

Yes, this is the current workflow. I explain it as follows: 

1. line chart -> switch to big number (do **transform** from `line` to `big number`)
2. big number -> switch to bar chart (do **transform** from `big number` to `bar chart`), the `big number` has not dimensions so the `bar chart` also has not dimensions.
3. if you do not do any operation on the `big number`, from `big number` switch to the `line chart`, `standardized form data` will do **restore**, so the line chart form_data can be restored.
---------------
- the *transform* means to transform from the previous chart(Whether you query the data or not.)
- the *restore* means to restore to the original form_data.

**Comment by kgabryje** [comment]
> Yes, this is the current workflow. I explain it as follows:
> 
> 1. line chart -> switch to big number (do **transform** from `line` to `big number`)
> 2. big number -> switch to bar chart (do **transform** from `big number` to `bar chart`), the `big number` has not dimensions so the `bar chart` also has not dimensions.
> 3. if you do not do any operation on the `big number`, from `big number` switch to the `line chart`, `standardized form data` will do **restore**, so the line chart form_data can be restored.
> 
> * the **transform** means to transform from the previous chart(Whether you query the data or not.)
> * the **restore** means to restore to the original form_data.
> 
> Right now, the Standardized Form Data doesn't provide a mechanism to "skip" a chart when do transformation.

I see your point! I'm still a bit afraid that this flow might be confusing for the end user who doesn't know how the feature works under the hood. I think more intuitive flow would be to keep controls in standardized form data and reset it only when user runs query. Like for example, if we saved dimensions in standardized form data while on line chart, they should still be there when user switches to big number, so they can carry over to next viz types. Only after user runs big number query the dimensions would get reset.
That's how I think I'd expect this feature to behave if I was the end user. Maybe we should get some input from product people? @lauderbaugh @rusackas what do you guys think?

**Comment by lauderbaugh** [comment]
I think an ideal flow might look something like:
If we have form data for `big number` and switch to `pie`, it will keep the form data. If we then switch to `line` it will keep the form data all the way from `big number`. If we make adjustments to `line` and then switch to `bar`, the adjustments from line will be reflected in `bar`, (though anything not adjusted would continue to bring the form data all the way from `big number` if possible). Does that make sense?

**Comment by kgabryje** [comment]
> I think an ideal flow might look something like: If we have form data for `big number` and switch to `pie`, it will keep the form data. If we then switch to `line` it will keep the form data all the way from `big number`. If we make adjustments to `line` and then switch to `bar`, the adjustments from line will be reflected in `bar`, (though anything not adjusted would continue to bring the form data all the way from `big number` if possible). Does that make sense?

The matter up for discussion is "should switching to Big Number (for example) reset the Dimensions in standardized form data". Big number doesn't have Dimensions control, while Line, Bar, Pie etc. do. So we need to decide what happens when we set open line chart, set a Dimension, then switch to Big Number, then switch to Bar. Should Bar chart have Dimension control populated with value set in Line chart, or should it be reset since the previous viz didn't have that control.

**Comment by rusackas** [comment]
> > I think an ideal flow might look something like: If we have form data for `big number` and switch to `pie`, it will keep the form data. If we then switch to `line` it will keep the form data all the way from `big number`. If we make adjustments to `line` and then switch to `bar`, the adjustments from line will be reflected in `bar`, (though anything not adjusted would continue to bring the form data all the way from `big number` if possible). Does that make sense?
> 
> The matter up for discussion is "should switching to Big Number (for example) reset the Dimensions in standardized form data". Big number doesn't have Dimensions control, while Line, Bar, Pie etc. do. So we need to decide what happens when we set open line chart, set a Dimension, then switch to Big Number, then switch to Bar. Should Bar chart have Dimension control populated with value set in Line chart, or should it be reset since the previous viz didn't have that control.

Thanks for distilling this discussion...

There might be a good reason not to do this (some danger I'm not clear on) but my gut feeling is this:

• All data that's modified or appended to the standardized controls should be persisted. You may switch to another viz that doesn't use it, but it should remain in memory/state. Then when you switch at any time to a viz looking for that control, it can be recalled. A field should not be disposed of until the chart is saved, and the ephemeral state is disposed of.

• I think the current logic that _already_ applies to other extraneous (non-standardized) controls (i.e. each chart type keeps its unique properties like a Pie/Donut's inner radius setting) is that that they are kept in that ephemeral state, and restored when returning to the Pie chart after any number of viz switches. Treating standardized controls the same way, so that they may always be restored, makes sense to me.

In other words, is there any harm in keeping that extra/unused data (Dimensions, in our case) in the ephemeral state, in case some other chart type finds it useful?

**Comment by kgabryje** [comment]
I noticed 1 bug, not sure if it was there before. When we switch from Big number to Table, we get a validation error, even though metrics control is filled. Deleting and re-adding the metric fixes the problem.

https://user-images.githubusercontent.com/15073128/174324033-6a9db929-98bc-4694-85ec-2e17035c19cb.mov

**Comment by zhaoyongjie** [comment]
> I noticed 1 bug, not sure if it was there before. When we switch from Big number to Table, we get a validation error, even though metrics control is filled. Deleting and re-adding the metric fixes the problem.
> Screen.Recording.2022-06-17.at.16.57.55.mov

big number sent a query, then form_data should follow big_numer's form_data. this means you want show a data layout is big_number, so dimension is lost.(the bignumber does not have dimension.)

**Comment by kgabryje** [comment]
> big number sent a query, then form_data should follow big_numer's form_data. this means you want show a data layout is big_number, so dimension is lost.(the bignumber does not have dimension.)
> 
> The `big_number` is one metric and 0 dimension viz, you can't get a dimension if you change to table or other more dimension viz.

I mean, table with metric added is correct form data, but the controls show errors (see video)

**Comment by zhaoyongjie** [comment]
> > big number sent a query, then form_data should follow big_numer's form_data. this means you want show a data layout is big_number, so dimension is lost.(the bignumber does not have dimension.)
> > The `big_number` is one metric and 0 dimension viz, you can't get a dimension if you change to table or other more dimension viz.
> 
> I mean, table with metric added is correct form data, but the controls show errors (see video)

@kgabryje Oh! I saw! Thanks for the heads up. this issue is also on the master branch. I will fix it in a separate PR.

**Comment by kgabryje** [comment]
Hmm the behaviour of reverting metrics is a bit weird... I think it shouldn't be reverted to viz type's memorized data? But take metrics/dimensions from standardized data?

https://user-images.githubusercontent.com/15073128/175029675-c9712947-4a6f-4b6a-bf3e-71854630119b.mov

**Comment by zhaoyongjie** [comment]
@kgabryje I think we've discussed this before. You don't have a query, so do a `restore` operation. not a `transform` operation. it's also you proposal for this feature.

**Comment by zhaoyongjie** [comment]
@kgabryje 
I wrote a script for your first footage. 

1. Line chart:  ['test metric']
do: transform -> bar
2. Bar chart: ['test metric']
do: edit, Bar chart, ['test metric', 'count(*)']
do: restore(non query) -> Line, because the bar chart in the memorizedFormData(stack)
3. Line chart: ['test metric']
do: transform -> bar
4. Bar chart: ['test metric'], transform from Line chart
do: edit, Bar chart, ['test metric', 'count(*)']
do: transform -> area, because the area chart not in the stack
5. Are chart: metrics: ['test metric', 'count(*)']
do: restore(non query) -> , because the Bar chart in the stack
6. Line chart: metrics: ['test metric']
....
....

**Review by kgabryje — APPROVED** [review]
LGTM! Thank you for your patience and this awesome feature 😄
